R/dist.R

```diff
@@ -1382,48 +1474,374 @@ plot.dist_spec <- function(x, ...) {
+          ) {
+          stop(
+            "Maximum of parameter ", param_name, " is less than its ",
+            "lower bound ", lb, "."
+          )
+        }
+      }
+
+      ## convert any unnatural parameters
+      unnatural_params <- setdiff(names(params), natural_params(distribution))
+      if (length(unnatural_params) > 0) {
+        if (length(unnatural_params) < length(params)) {
+          stop(
+            "Incompatible combination of parameters of a ", distribution,
+            " distribution specified."
+          )
+        }
+        ## sample parameters if they are uncertain
+        if (any(vapply(params, sd_dist, numeric(1)) > 0)) {
+          warning(
+            "Uncertain ", distribution, " distribution specified in terms of ",
+            "parameters that are not the \"natural\" parameters of the ",
+            "distribution (", toString(natural_params(distribution)),
+            "). Converting using a crude and very approximate method ",
+            "that is likely to produce biased results. If possible, ",
+            "it is preferable to specify the distribution directly ",
+            "in terms of the natural parameters."
+          )
+        }
+        ## generate natural parameters
+        params <- convert_to_natural(params, distribution)
+      }
+      ## convert normal with sd == 0 to fixed
+      if (distribution == "normal" && is.numeric(params$sd) && params$sd == 0) {
+        ret <- list(
+          parameters = list(value = params$mean), distribution = "fixed"
+        )
+      } else {
+        ret <- list(parameters = params, distribution = distribution)
+      }
+    }
+    ret <- c(ret, list(max = max))
   }
-  return(x)
+  ## join and wrap in another list to make concatenating easier
+  ret <- list(ret)
+  attr(ret, "class") <- c("dist_spec", "list")
+
+  ## now we have a distribution with natural parameters - return dist_spec
+  return(ret)
+}
+
+##' Internal function for converting parameters to natural parameters.
+##'
+##' This is used for preprocessing before generating a `dist_spec` object
+##' from a given set of parameters and distribution
+##' @param params A numerical named parameter vector
+##' @inheritParams natural_params
+##' @return A list with two elements, `params_mean` and `params_sd`, containing
+##' mean and sd of natural parameters.
+##' @author Sebastian Funk
+convert_to_natural <- function(params, distribution) {
+  ## unnatural parameter means
+  ux <- lapply(params, mean)
+  ## estimate relative uncertainty of parameters
+  rel_unc <- mean(vapply(params, sd_dist, numeric(1))^2 / unlist(ux))
+  ## store natural parameters
+  x <- list()
+  if (distribution == "gamma") {
+    if ("mean" %in% names(ux) && "sd" %in% names(ux)) {
+      x$shape <- ux$mean**2 / ux$sd**2
+      x$rate <- x$shape / ux$mean
+    } else if (!("rate" %in% names(ux)) && ("scale" %in% names(ux))) {
+      x$shape <- ux$shape
+      x$rate <- 1 / ux$scale
```

Comment:
This looks like it should work but the test gap means it might not. I'd suggest add a test for the gamma along the lines of the example I suggested and also adding a unit test here